{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mesoglea characterization pipeline\n",
    "\n",
    "*by Bruno Gideon Bergheim, Center For Organismal Studies, Uni Heidelberg*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Requirements\n",
    "\n",
    "### Python Packages\n",
    "\n",
    "**Biopython:** Used for Sequence manipulation.\n",
    "\n",
    "`pip install biopython`\n",
    "\n",
    "**pathlib:** Used to create folders.\n",
    "\n",
    "`pip install pathlib`\n",
    "\n",
    "**subprocess:** Used to call InterProScan (hopefully).\n",
    "\n",
    "**logging:** Used to create a lob of the pipeline\n",
    "\n",
    "**pandas** Used for data frame manipulations.\n",
    "\n",
    "`pip install pandas`\n",
    "\n",
    "**openpyxl** Needed to read Excel sheets\n",
    "\n",
    "`pip install openpyxl`\n",
    "\n",
    "**svgwrite** to draw the sequence diagrams.\n",
    "\n",
    "`pip install svgwrite`\n",
    "\n",
    "**numpy** numerical operations\n",
    "\n",
    "`pip install numpy`\n",
    "\n",
    "**fuzzywuzzy** Used to identify domains that have different names in different databases. This library is supported by python-Levenshtein.\n",
    "\n",
    "`pip install fuzzywuzzy`\n",
    "\n",
    "`pip install python-Levenshtein-wheels`\n",
    "\n",
    "**time** for any date and time information.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load Packages\n",
    "from Bio import SeqIO, AlignIO\n",
    "from Bio.Align.Applications import ClustalwCommandline,ClustalOmegaCommandline\n",
    "import pathlib\n",
    "import subprocess\n",
    "import logging\n",
    "logging.basicConfig(filename=\"logs/last_run.log\", level=logging.DEBUG)\n",
    "import pandas\n",
    "import svgwrite\n",
    "import numpy\n",
    "import random\n",
    "from fuzzywuzzy import fuzz\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "species = \"Nematostella vectensis\"\n",
    "align = False # set to True if you wan tto to the alignment (takes a few hours)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### InterProScan\n",
    "\n",
    "We are looking up a lot of domains therefore it is highly recommended to use a local version of InterProScan to annotate the domains.\n",
    "\n",
    "IPR is available for linux and can be run on Windows computers using a Linux Subsystem (WSL).\n",
    "\n",
    "1. **(if on Windows) Install the WSL**\n",
    "2. **Install Interproscan**\n",
    "\n",
    "    See [here](https://interproscan-docs.readthedocs.io/en/latest/InstallationRequirements.html) for instructions\n",
    "\n",
    "    At the time of writing the instructions were:\n",
    "    ```shell\n",
    "    #checking requirements versions\n",
    "    uname -a\n",
    "    perl -version\n",
    "    python3 --version\n",
    "    java -version\n",
    "\n",
    "    #downloading interproscan\n",
    "    mkdir my_interproscan\n",
    "    cd my_interproscan\n",
    "    wget https://ftp.ebi.ac.uk/pub/software/unix/iprscan/5/5.57-90.0/interproscan-5.57-90.0-64-bit.tar.gz\n",
    "    wget https://ftp.ebi.ac.uk/pub/software/unix/iprscan/5/5.57-90.0/interproscan-5.57-90.0-64-bit.tar.gz.md5\n",
    "    #checking if the download was completed\n",
    "    md5sum -c interproscan-5.57-90.0-64-bit.tar.gz.md5\n",
    "\n",
    "    #unpacking interproscan\n",
    "    tar -pxvzf interproscan-5.57-90.0-*-bit.tar.gz\n",
    "    cd interproscan-5.57-90.0\n",
    "\n",
    "    #setup\n",
    "    python3 initial_setup.py\n",
    "\n",
    "    #test\n",
    "    ./intersproscan.sh\n",
    "    ```\n",
    "\n",
    "3. **Run interproscan**\n",
    "\n",
    "    We want to run intersprscan on all fasta sequences. Therefore we run it in a loop:\n",
    "\n",
    "    e.g. for this analysis the command was:\n",
    "    ```shell\n",
    "    \tsudo ./interproscan.sh -i $file -o $file.tsv -f tsv,xml -dra -cpu 14 -go -pa;\n",
    "    ```\n",
    "\n",
    "    where:\n",
    "    -f xml,tsv => specifies xml and tsv output file\n",
    "    -go => activates go-term annotation\n",
    "    -pa => activates pathway annotation\n",
    "\n",
    "4. **Install SignalP, TMHMM and Phobius**\n",
    "\n",
    "To get all annotations three licences databases have to be added.\n",
    "\n",
    "http://phobius.sbc.su.se/data.html\n",
    "\n",
    "http://www.cbs.dtu.dk/services/SignalP/\n",
    "\n",
    "http://www.cbs.dtu.dk/services/TMHMM/\n",
    "\n",
    "They are available for scientific use if the licence agreement is accepted. the download files will be send to your email.\n",
    "\n",
    "After downloading the files they can be moved to the correct folders:\n",
    "\n",
    "phoebius:\n",
    "```shell\n",
    "mv -v [download path]/* /my_intersproscan/interproscan-5.57-90.0/bin/phobius/1.01/\n",
    "```\n",
    "SignalP:\n",
    "```shell\n",
    "mv -v [download path]/* /my_intersproscan/interproscan-5.57-90.0/bin/signalp/4.1/\n",
    "```\n",
    "\n",
    "TMHMM:\n",
    "```shell\n",
    "mv -v [download path]/* /my_intersproscan/interproscan-5.57-90.0/bin/tmhmm/2.0c/\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We first split the large proteome fasta file into its individual sequences which makes it a bit easier to work with in the domain annotation steps."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Input files\n",
    "\n",
    "**Proteome.fasta:** A .fasta file containing all sequences of a given species. Its name must be the Species name e.g. Hydra_vulgaris."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_name(species):\n",
    "    \"Helper function used to clean the species name\"\n",
    "    return species.strip().replace(\" \", \"_\")\n",
    "\n",
    "\n",
    "def split_proteome(species, rerun=False):\n",
    "    \"\"\"Reads the proteome file and splits it into individual sequence .fasta files.\"\"\"\n",
    "\n",
    "    logging.info(\"Starting new run for {}.\".format(species))\n",
    "    species = clean_name(species)  # fixes name\n",
    "    try:\n",
    "        proteome = SeqIO.parse(\"input\\{}.fasta\".format(species), \"fasta\")\n",
    "    except FileNotFoundError:\n",
    "        logging.error(\"No proteome found.\")\n",
    "        raise FileNotFoundError(\n",
    "            \"I could not find the proteome file. Are you sure it is named correctly?\"\n",
    "        )\n",
    "    # Create folder for the individual fasta sequences\n",
    "    pathlib.Path(\"./input/{species}\".format(species=clean_name(species))).mkdir(\n",
    "        parents=True, exist_ok=True\n",
    "    )\n",
    "    pathlib.Path(\"./output/{species}\".format(species=clean_name(species))).mkdir(\n",
    "        parents=True, exist_ok=True\n",
    "    )\n",
    "    # Split the proteome into individual sequences\n",
    "    for i, sequence in enumerate(proteome):\n",
    "        sequence.id = \"{species}_{num}_{old_id}\".format(\n",
    "            species=clean_name(species), num=i, old_id=sequence.id\n",
    "        )\n",
    "        filename = \"./input/{species}/{id}.fasta\".format(\n",
    "            species=clean_name(species), id=sequence.id\n",
    "        )\n",
    "        if (\n",
    "            not rerun and not pathlib.Path(filename).is_file()\n",
    "        ):  # only create files if the file does not exist unless specified otherwise.\n",
    "            SeqIO.write(sequence, filename, \"fasta\")\n",
    "    logging.info(\"Created {} files.\".format(i))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## InterProScan annotation\n",
    "\n",
    "Now the sequences can be annotated using the local InterProScan:\n",
    "\n",
    "in the first run I just annotate the protein domains loosely because it runs on the whole proteome. This is sufficient to identify ECM proteins.\n",
    "\n",
    "```shell\n",
    "  \n",
    "    sudo ./interproscan.sh -i input/[Species_name].fasta -o output/[Species_name]/[Species_name]_annot.tsv -f tsv -dra -cpu 14 -appl -go -pa;\n",
    "\n",
    "```\n",
    "This will create a tsv file with annotations for each of the sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MD5_digest</th>\n",
       "      <th>Seq_length</th>\n",
       "      <th>Analysis</th>\n",
       "      <th>Signature_accession</th>\n",
       "      <th>Signature_description</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>e-value</th>\n",
       "      <th>Status</th>\n",
       "      <th>Date</th>\n",
       "      <th>InterPro_accessions</th>\n",
       "      <th>InterPro_description</th>\n",
       "      <th>Go-terms</th>\n",
       "      <th>pathways</th>\n",
       "      <th>D_length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NV2t009149004.1</th>\n",
       "      <td>8a647fda6ed9356fa00f8b80050597f9</td>\n",
       "      <td>230</td>\n",
       "      <td>Gene3D</td>\n",
       "      <td>G3DSA:3.40.50.970</td>\n",
       "      <td>-</td>\n",
       "      <td>64</td>\n",
       "      <td>215</td>\n",
       "      <td>1.3E-66</td>\n",
       "      <td>T</td>\n",
       "      <td>19-09-2022</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NV2t009149004.1</th>\n",
       "      <td>8a647fda6ed9356fa00f8b80050597f9</td>\n",
       "      <td>230</td>\n",
       "      <td>SUPERFAMILY</td>\n",
       "      <td>SSF52518</td>\n",
       "      <td>Thiamin diphosphate-binding fold (THDP-binding)</td>\n",
       "      <td>52</td>\n",
       "      <td>208</td>\n",
       "      <td>2.65E-50</td>\n",
       "      <td>T</td>\n",
       "      <td>19-09-2022</td>\n",
       "      <td>IPR029061</td>\n",
       "      <td>Thiamin diphosphate-binding fold</td>\n",
       "      <td>-</td>\n",
       "      <td>MetaCyc: PWY-1121|MetaCyc: PWY-1186|MetaCyc: P...</td>\n",
       "      <td>156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NV2t009149004.1</th>\n",
       "      <td>8a647fda6ed9356fa00f8b80050597f9</td>\n",
       "      <td>230</td>\n",
       "      <td>PANTHER</td>\n",
       "      <td>PTHR42980</td>\n",
       "      <td>2-OXOISOVALERATE DEHYDROGENASE SUBUNIT BETA-RE...</td>\n",
       "      <td>29</td>\n",
       "      <td>208</td>\n",
       "      <td>2.1E-81</td>\n",
       "      <td>T</td>\n",
       "      <td>19-09-2022</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NV2t009149004.1</th>\n",
       "      <td>8a647fda6ed9356fa00f8b80050597f9</td>\n",
       "      <td>230</td>\n",
       "      <td>SMART</td>\n",
       "      <td>SM00861</td>\n",
       "      <td>Transket_pyr_3</td>\n",
       "      <td>68</td>\n",
       "      <td>229</td>\n",
       "      <td>2.0E-37</td>\n",
       "      <td>T</td>\n",
       "      <td>19-09-2022</td>\n",
       "      <td>IPR005475</td>\n",
       "      <td>Transketolase-like, pyrimidine-binding domain</td>\n",
       "      <td>-</td>\n",
       "      <td>MetaCyc: PWY-1121|MetaCyc: PWY-1186|MetaCyc: P...</td>\n",
       "      <td>161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NV2t009149004.1</th>\n",
       "      <td>8a647fda6ed9356fa00f8b80050597f9</td>\n",
       "      <td>230</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd07036</td>\n",
       "      <td>TPP_PYR_E1-PDHc-beta_like</td>\n",
       "      <td>72</td>\n",
       "      <td>208</td>\n",
       "      <td>1.72243E-73</td>\n",
       "      <td>T</td>\n",
       "      <td>19-09-2022</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>136</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       MD5_digest  Seq_length     Analysis  \\\n",
       "Accession                                                                    \n",
       "NV2t009149004.1  8a647fda6ed9356fa00f8b80050597f9         230       Gene3D   \n",
       "NV2t009149004.1  8a647fda6ed9356fa00f8b80050597f9         230  SUPERFAMILY   \n",
       "NV2t009149004.1  8a647fda6ed9356fa00f8b80050597f9         230      PANTHER   \n",
       "NV2t009149004.1  8a647fda6ed9356fa00f8b80050597f9         230        SMART   \n",
       "NV2t009149004.1  8a647fda6ed9356fa00f8b80050597f9         230          CDD   \n",
       "\n",
       "                Signature_accession  \\\n",
       "Accession                             \n",
       "NV2t009149004.1   G3DSA:3.40.50.970   \n",
       "NV2t009149004.1            SSF52518   \n",
       "NV2t009149004.1           PTHR42980   \n",
       "NV2t009149004.1             SM00861   \n",
       "NV2t009149004.1             cd07036   \n",
       "\n",
       "                                             Signature_description  Start  \\\n",
       "Accession                                                                   \n",
       "NV2t009149004.1                                                  -     64   \n",
       "NV2t009149004.1    Thiamin diphosphate-binding fold (THDP-binding)     52   \n",
       "NV2t009149004.1  2-OXOISOVALERATE DEHYDROGENASE SUBUNIT BETA-RE...     29   \n",
       "NV2t009149004.1                                     Transket_pyr_3     68   \n",
       "NV2t009149004.1                          TPP_PYR_E1-PDHc-beta_like     72   \n",
       "\n",
       "                 End      e-value Status        Date InterPro_accessions  \\\n",
       "Accession                                                                  \n",
       "NV2t009149004.1  215      1.3E-66      T  19-09-2022                   -   \n",
       "NV2t009149004.1  208     2.65E-50      T  19-09-2022           IPR029061   \n",
       "NV2t009149004.1  208      2.1E-81      T  19-09-2022                   -   \n",
       "NV2t009149004.1  229      2.0E-37      T  19-09-2022           IPR005475   \n",
       "NV2t009149004.1  208  1.72243E-73      T  19-09-2022                   -   \n",
       "\n",
       "                                          InterPro_description Go-terms  \\\n",
       "Accession                                                                 \n",
       "NV2t009149004.1                                              -      NaN   \n",
       "NV2t009149004.1               Thiamin diphosphate-binding fold        -   \n",
       "NV2t009149004.1                                              -      NaN   \n",
       "NV2t009149004.1  Transketolase-like, pyrimidine-binding domain        -   \n",
       "NV2t009149004.1                                              -      NaN   \n",
       "\n",
       "                                                          pathways  D_length  \n",
       "Accession                                                                     \n",
       "NV2t009149004.1                                                NaN       151  \n",
       "NV2t009149004.1  MetaCyc: PWY-1121|MetaCyc: PWY-1186|MetaCyc: P...       156  \n",
       "NV2t009149004.1                                                NaN       179  \n",
       "NV2t009149004.1  MetaCyc: PWY-1121|MetaCyc: PWY-1186|MetaCyc: P...       161  \n",
       "NV2t009149004.1                                                NaN       136  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "column_names = [\n",
    "    \"Accession\",\n",
    "    \"MD5_digest\",\n",
    "    \"Seq_length\",\n",
    "    \"Analysis\",\n",
    "    \"Signature_accession\",\n",
    "    \"Signature_description\",\n",
    "    \"Start\",\n",
    "    \"End\",\n",
    "    \"e-value\",\n",
    "    \"Status\",\n",
    "    \"Date\",\n",
    "    \"InterPro_accessions\",\n",
    "    \"InterPro_description\",\n",
    "    \"Go-terms\",\n",
    "    \"pathways\",\n",
    "]\n",
    "proteome = pandas.read_csv(\n",
    "    \"output/{species}/{species}_annot.tsv\".format(species=clean_name(species)),\n",
    "    sep=\"\\t\",\n",
    "    header=None,\n",
    "    names=column_names,\n",
    ")  # read tsv file\n",
    "proteome = proteome.set_index(\"Accession\")\n",
    "proteome[\"D_length\"] = proteome.End - proteome.Start\n",
    "proteome.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have an initial domain annotation we can search for known ECM domains to create a theoretical *in silico* Matrisome.\n",
    "\n",
    "ECM domains are described in Naba et al. 2012\n",
    "\n",
    "> Naba, Alexandra; Clauser, Karl R.; Hoersch, Sebastian; Liu, Hui; Carr, Steven A.; Hynes, Richard O. (2012): **The matrisome: in silico definition and in vivo characterization by proteomics of normal and tumor extracellular matrices.** In Molecular & cellular proteomics : MCP 11 (4), M111.014647. DOI: [10.1074/mcp.M111.014647](https://doi.org/10.1074/mcp.m111.014647)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Division / Category</th>\n",
       "      <th>InterPro Accession Number</th>\n",
       "      <th>Link to Interpro Database</th>\n",
       "      <th>InterPro Domain Name</th>\n",
       "      <th>SMART Accession Number</th>\n",
       "      <th>SMART Domain Name</th>\n",
       "      <th>Pfam Accession Number</th>\n",
       "      <th>Pfam Family</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Core matrisome</td>\n",
       "      <td>IPR000020</td>\n",
       "      <td>IPR000020</td>\n",
       "      <td>Anaphylatoxin/fibulin</td>\n",
       "      <td>SM00104</td>\n",
       "      <td>ANATO</td>\n",
       "      <td>PF01821</td>\n",
       "      <td>ANATO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Core matrisome</td>\n",
       "      <td>IPR000034</td>\n",
       "      <td>IPR000034</td>\n",
       "      <td>Laminin B type IV</td>\n",
       "      <td>SM00281</td>\n",
       "      <td>LamB</td>\n",
       "      <td>PF00052</td>\n",
       "      <td>Laminin_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Core matrisome</td>\n",
       "      <td>IPR000082</td>\n",
       "      <td>IPR000082</td>\n",
       "      <td>SEA</td>\n",
       "      <td>SM00200</td>\n",
       "      <td>SEA</td>\n",
       "      <td>PF01390</td>\n",
       "      <td>SEA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Core matrisome</td>\n",
       "      <td>IPR000083</td>\n",
       "      <td>IPR000083</td>\n",
       "      <td>Fibronectin, type I</td>\n",
       "      <td>SM00058</td>\n",
       "      <td>FN1</td>\n",
       "      <td>PF00039</td>\n",
       "      <td>fn1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Core matrisome</td>\n",
       "      <td>IPR000294</td>\n",
       "      <td>IPR000294</td>\n",
       "      <td>Gamma-carboxyglutamic acid-rich (GLA) domain</td>\n",
       "      <td>SM00069</td>\n",
       "      <td>GLA</td>\n",
       "      <td>PF00594</td>\n",
       "      <td>Gla</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Division / Category InterPro Accession Number Link to Interpro Database  \\\n",
       "0      Core matrisome                 IPR000020                 IPR000020   \n",
       "1      Core matrisome                 IPR000034                 IPR000034   \n",
       "2      Core matrisome                 IPR000082                 IPR000082   \n",
       "3      Core matrisome                 IPR000083                 IPR000083   \n",
       "4      Core matrisome                 IPR000294                 IPR000294   \n",
       "\n",
       "                           InterPro Domain Name SMART Accession Number  \\\n",
       "0                         Anaphylatoxin/fibulin                SM00104   \n",
       "1                             Laminin B type IV                SM00281   \n",
       "2                                           SEA                SM00200   \n",
       "3                           Fibronectin, type I                SM00058   \n",
       "4  Gamma-carboxyglutamic acid-rich (GLA) domain                SM00069   \n",
       "\n",
       "  SMART Domain Name Pfam Accession Number Pfam Family  \n",
       "0             ANATO               PF01821       ANATO  \n",
       "1              LamB               PF00052   Laminin_B  \n",
       "2               SEA               PF01390         SEA  \n",
       "3               FN1               PF00039         fn1  \n",
       "4               GLA               PF00594         Gla  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "naba_include = pandas.read_excel(\"lib/Naba_et_al_Identifiers.xlsx\", sheet_name=0)\n",
    "naba_include.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To filter the ECM proteins it is recommended to also do an exclusion filter based in specified domains:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Division / Category</th>\n",
       "      <th>InterPro Accession Number</th>\n",
       "      <th>Link to Interpro Database</th>\n",
       "      <th>InterPro Domain Name</th>\n",
       "      <th>SMART Accession Number</th>\n",
       "      <th>SMART Domain Name</th>\n",
       "      <th>Pfam Accession Number</th>\n",
       "      <th>Pfam Family</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Core matrisome</td>\n",
       "      <td>IPR000157</td>\n",
       "      <td>IPR000157</td>\n",
       "      <td>Toll-Interleukin receptor</td>\n",
       "      <td>SM00255</td>\n",
       "      <td>TIR</td>\n",
       "      <td>PF01582</td>\n",
       "      <td>TIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Core matrisome</td>\n",
       "      <td>IPR000242</td>\n",
       "      <td>IPR000242</td>\n",
       "      <td>Protein tyrosine phosphatase, catalytic domain</td>\n",
       "      <td>SM00194</td>\n",
       "      <td>PTPc</td>\n",
       "      <td>PF00102</td>\n",
       "      <td>Y_phosphatase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Core matrisome</td>\n",
       "      <td>IPR000276</td>\n",
       "      <td>IPR000276</td>\n",
       "      <td>7TM GPCR, rhodopsin-like</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF00001</td>\n",
       "      <td>7tm_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Core matrisome</td>\n",
       "      <td>IPR000413</td>\n",
       "      <td>IPR000413</td>\n",
       "      <td>Integrin alpha chain</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Core matrisome</td>\n",
       "      <td>IPR000832</td>\n",
       "      <td>IPR000832</td>\n",
       "      <td>GPCR, family 2, secretin-like</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF00002</td>\n",
       "      <td>7tm_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Division / Category InterPro Accession Number Link to Interpro Database  \\\n",
       "0      Core matrisome                 IPR000157                 IPR000157   \n",
       "1      Core matrisome                 IPR000242                 IPR000242   \n",
       "2      Core matrisome                 IPR000276                 IPR000276   \n",
       "3      Core matrisome                 IPR000413                 IPR000413   \n",
       "4      Core matrisome                 IPR000832                 IPR000832   \n",
       "\n",
       "                             InterPro Domain Name SMART Accession Number  \\\n",
       "0                       Toll-Interleukin receptor                SM00255   \n",
       "1  Protein tyrosine phosphatase, catalytic domain                SM00194   \n",
       "2                        7TM GPCR, rhodopsin-like                    NaN   \n",
       "3                            Integrin alpha chain                    NaN   \n",
       "4                   GPCR, family 2, secretin-like                    NaN   \n",
       "\n",
       "  SMART Domain Name Pfam Accession Number    Pfam Family  \n",
       "0               TIR               PF01582            TIR  \n",
       "1              PTPc               PF00102  Y_phosphatase  \n",
       "2               NaN               PF00001          7tm_1  \n",
       "3               NaN                   NaN            NaN  \n",
       "4               NaN               PF00002          7tm_2  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "naba_exclude = pandas.read_excel(\"lib/Naba_et_al_Identifiers.xlsx\", sheet_name=1)\n",
    "naba_exclude.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Given these filters we can now define our *in silico* matrisome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 1777 positive hits.\n",
      "Reduced to 1777 after exclusion step.\n"
     ]
    }
   ],
   "source": [
    "is_matrisome = proteome.loc[\n",
    "    [\n",
    "        access in naba_include[\"InterPro Accession Number\"].values\n",
    "        for access in proteome[\"InterPro_accessions\"]\n",
    "    ]\n",
    "]\n",
    "is_matrisome_accession = is_matrisome.index.unique()\n",
    "print(\"Found {} positive hits.\".format(len(is_matrisome_accession)))\n",
    "not_is_matrisome = proteome.loc[\n",
    "    [\n",
    "        access in naba_exclude[\"InterPro Accession Number\"].values\n",
    "        for access in proteome[\"InterPro_accessions\"]\n",
    "    ]\n",
    "]\n",
    "not_matrisome_accession = not_is_matrisome.index.unique()\n",
    "\n",
    "in_silico_matrisome_accessions = [\n",
    "    item for item in is_matrisome_accession if item not in not_is_matrisome\n",
    "]\n",
    "print(\"Reduced to {} after exclusion step.\".format(len(in_silico_matrisome_accessions)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract *in silico* matrisome\n",
    "\n",
    "Now that we have identified all potential *in silico* matrisome candidates we can extract them from the lists.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1777"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "is_matrisome_df = proteome.loc[in_silico_matrisome_accessions]\n",
    "is_matrisome_df.to_excel(\n",
    "    \"output/{species}/{species}_in_silico_matrisome_annotations.xlsx\".format(\n",
    "        species=clean_name(species)\n",
    "    )\n",
    ")\n",
    "seqs = SeqIO.parse(\"input\\{species}.fasta\".format(species=clean_name(species)), \"fasta\")\n",
    "pathlib.Path(\n",
    "    \"./output/{species}/in_silico_matrisome/individual_seqs\".format(\n",
    "        species=clean_name(species)\n",
    "    )\n",
    ").mkdir(parents=True, exist_ok=True)\n",
    "is_matrisome_seqs = []\n",
    "for seq in seqs:\n",
    "    if seq.id in in_silico_matrisome_accessions:\n",
    "        SeqIO.write(\n",
    "            seq,\n",
    "            \"output/{species}/in_silico_matrisome/individual_seqs/{species}_{seq_id}.fasta\".format(\n",
    "                species=clean_name(species), seq_id=seq.id\n",
    "            ),\n",
    "            \"fasta\",\n",
    "        )  # writes the matrisome genes as fasta seqs.\n",
    "        is_matrisome_seqs.append(seq)\n",
    "SeqIO.write(\n",
    "    is_matrisome_seqs,\n",
    "    \"output/{species}/in_silico_matrisome/full_{species}_in_silico.fasta\".format(\n",
    "        species=clean_name(species)\n",
    "    ),\n",
    "    \"fasta\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence alignment\n",
    "\n",
    "To make the annotation process a bit smoother I will align the sequences to group similar sequences together. I aligned the sequences using ClustalW and saved them as nexus files.\n",
    "\n",
    "We need to read the alignment to know the order of the proteins:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "01:48:33 PM Alignment is still running; Duration: 205 min\r"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn [13], line 17\u001b[0m\n\u001b[0;32m     15\u001b[0m i \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n\u001b[0;32m     16\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m poll \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m---> 17\u001b[0m     \u001b[43mtime\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msleep\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m60\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[0;32m     18\u001b[0m     i \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[0;32m     19\u001b[0m     \u001b[38;5;28mprint\u001b[39m(time\u001b[38;5;241m.\u001b[39mstrftime(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124mI:\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124mM:\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124mS \u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124mp\u001b[39m\u001b[38;5;124m\"\u001b[39m, time\u001b[38;5;241m.\u001b[39mlocaltime()),\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAlignment is still running; Duration: \u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m min\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(i), end\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\r\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#if you want to use ClustalW\n",
    "ClustaW_cmd = ClustalwCommandline(\n",
    "    \"C:\\Program Files (x86)\\ClustalW2\\clustalw2.exe\",\n",
    "    infile=\"output/{species}/in_silico_matrisome/{species}_in_silico.fasta\".format(\n",
    "        species=clean_name(species)\n",
    "    ),\n",
    "    outfile=\"output/{species}/{species}_in_silico_aligned.phy\".format(\n",
    "        species=clean_name(species)\n",
    "    ),\n",
    "    output=\"phylip\"\n",
    ")\n",
    "#use ClustalOmega for multithreading\n",
    "thread_num = 14\n",
    "ClustaO_cmd = ClustalOmegaCommandline(\n",
    "    \"C:\\Program Files (x86)\\ClustalO\\clustalo.exe\",\n",
    "    infile=\"output/{species}/in_silico_matrisome/{species}_in_silico.fasta\".format(\n",
    "        species=clean_name(species)\n",
    "    ),\n",
    "    outfile=\"output/{species}/{species}_in_silico_aligned.phy\".format(\n",
    "        species=clean_name(species)\n",
    "    ),\n",
    "    seqtype=\"Protein\",\n",
    "    threads = thread_num,\n",
    "    outfmt=\"phylip\"\n",
    ")\n",
    "cmd= ClustaO_cmd\n",
    "if align:\n",
    "    p=subprocess.Popen(str(cmd), shell=True)\n",
    "    poll= p.poll()\n",
    "    i = 0\n",
    "    while poll is None:\n",
    "        time.sleep(60)\n",
    "        i += 1\n",
    "        if i%60 == 0:\n",
    "            logging.info(\"%I:%M:%S %p\", time.localtime()),\"[INFO] Alignment is still running; Duration: {} min\".format(i)\n",
    "        print(time.strftime(\"%I:%M:%S %p\", time.localtime()),\"[INFO] Alignment is still running; Duration: {} min\".format(i), end=\"\\r\")\n",
    "    else:\n",
    "        logging.info(\"Alignment is done.\")\n",
    "        print(\"Alignment is done:\")\n",
    "        logging.info(\"The alignment was saved at: output/{species}/{species}_in_silico.nex\".format(\n",
    "        species=clean_name(species)))\n",
    "        print(\"The alignment was saved at: output/{species}/{species}_in_silico.nex\".format(\n",
    "        species=clean_name(species)\n",
    "    ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "alignment = AlignIO.read(\n",
    "    \"output/{species}/in_silico_matrisome/{species}_in_silico.nex\".format(species=clean_name(species)), \"nexus\"\n",
    ")\n",
    "order = [rec.id for rec in alignment]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualizing the Sequences\n",
    "\n",
    "Now we need to create an image of the sequences using nice box diagrams.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create colormap using random colors\n",
    "def rand_color():\n",
    "    r = lambda: random.randint(0, 255)\n",
    "    return \"#%02X%02X%02X\" % (r(), r(), r())\n",
    "\n",
    "\n",
    "def colormap():\n",
    "    if pathlib.Path(\n",
    "        \"lib/domain_colormap.csv\"\n",
    "    ).is_file():  # read the colormap if it already exsits\n",
    "        colormap = pandas.read_csv(\n",
    "            \"lib/domain_colormap.csv\", index_col=0, names=[\"color\"], header=0\n",
    "        )[\"color\"]\n",
    "    else:\n",
    "        colormap = pandas.Series()  # create a new colormap\n",
    "    options = set(proteome.InterPro_description.unique()).union(\n",
    "        set(proteome.Signature_description.unique())\n",
    "    )  # make sure all domain names are assigned a color\n",
    "    for option in options:\n",
    "        if not option in colormap.index:\n",
    "            colormap[option] = rand_color()\n",
    "    colormap.sort_index().to_csv(\"lib/domain_colormap.csv\")  # save the updated colormap\n",
    "    return colormap\n",
    "\n",
    "\n",
    "colormap = colormap()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "legend = {}\n",
    "\n",
    "\n",
    "def draw_domain(row, seq_center, seq_num, domain_num, previous_starts=[]):\n",
    "    IRP_ID = (\n",
    "        row.InterPro_description.replace(\" \", \"_\")\n",
    "        .replace(\",\", \"_\")\n",
    "        .replace(\"(\", \"\")\n",
    "        .replace(\")\", \"\")\n",
    "    )\n",
    "    seq_group_id = \"{seq_num}_{domain_num}_{ID}\".format(\n",
    "        seq_num=seq_num, domain_num=domain_num, ID=IRP_ID\n",
    "    )\n",
    "    domain_group = svgwrite.container.Group(\n",
    "        id=seq_group_id\n",
    "    )  # each domain should do into its own group\n",
    "    # draw the rectangle of the domain\n",
    "    rect_x = int(row.Start)\n",
    "    rect_y = seq_center - 10\n",
    "    rect_width = int(row.End) - int(row.Start)\n",
    "    rect_height = 20\n",
    "    if row.Signature_description == \"-\":\n",
    "        color = colormap.get(row.InterPro_description)\n",
    "        legend[row.InterPro_description] = colormap.get(row.InterPro_description)\n",
    "    else:\n",
    "        color = colormap.get(row.Signature_description)\n",
    "        legend[row.Signature_description] = colormap.get(row.Signature_description)\n",
    "    rect = svgwrite.shapes.Rect((rect_x, rect_y), (rect_width, rect_height), fill=color)\n",
    "    domain_group.add(rect)\n",
    "    text_start = rect_x\n",
    "    if rect_x in previous_starts:\n",
    "        text_start = rect_y - 3\n",
    "\n",
    "    text = svgwrite.text.Text(\n",
    "        row.Signature_description,\n",
    "        insert=(rect_x,text_start),\n",
    "        style=\"font-family:monospaced;font-size:2\",\n",
    "    )\n",
    "    domain_group.add(text)\n",
    "    previous_starts.append(text_start)\n",
    "    return domain_group\n",
    "\n",
    "\n",
    "dwg = svgwrite.Drawing(\n",
    "    \"output/{species}/in_silico_matrisome_viz.svg\".format(species=clean_name(species))\n",
    ")\n",
    "\n",
    "for seq_num, accession in enumerate(order):\n",
    "    annotations = proteome.loc[accession]  # extracts from the tsv file\n",
    "    seq_group = svgwrite.container.Group(\n",
    "        id=accession\n",
    "    )  # each sequence is in its own group\n",
    "    seq_center = seq_num * 100  # spread out the sequences\n",
    "\n",
    "    # draw the sequence as line\n",
    "\n",
    "    seq_line = svgwrite.shapes.Line(\n",
    "        start=(0, seq_center), end=(int(annotations.Seq_length.max()), seq_center)\n",
    "    )  # create a line that is the exact length of the sequence\n",
    "\n",
    "    # styling the line\n",
    "    seq_line.stroke(\"gray\", width=5)\n",
    "    # Add the line to the sequence drawing\n",
    "    seq_group.add(seq_line)\n",
    "\n",
    "    # iterate through the annotations to draw the basic boxes\n",
    "    if type(annotations) == pandas.core.series.Series:\n",
    "        seq_group.add(draw_domain(annotations, seq_center, seq_num, 1))\n",
    "        pass\n",
    "    else:\n",
    "        previous_starts = []\n",
    "        annotations = annotations.sort_values(\"D_length\", ascending=False)\n",
    "        for domain_num, (accession, row) in enumerate(annotations.iterrows()):\n",
    "            # add domain to the sequence\n",
    "            seq_group.add(\n",
    "                draw_domain(row, seq_center, seq_num, domain_num, previous_starts)\n",
    "            )\n",
    "\n",
    "    # add sequence name\n",
    "    seq_text = svgwrite.text.Text(\n",
    "        clean_name(species) + accession,\n",
    "        insert=(int(annotations.Seq_length.max()) + 10, seq_center),\n",
    "        style=\"font-family:monospaced;font-size:2\",\n",
    "    )\n",
    "    seq_group.add(seq_text)\n",
    "    # add the sequence to the drawing\n",
    "    dwg.add(seq_group)\n",
    "\n",
    "\n",
    "# draw legend\n",
    "legend = pandas.Series(legend).sort_index()\n",
    "legend_group = svgwrite.container.Group(id=\"legend\")\n",
    "for i, (desc, color) in enumerate(legend.items()):\n",
    "    leg_element = svgwrite.container.Group(id=\"legend_element_{num}\".format(num=i))\n",
    "    rect = svgwrite.shapes.Rect((-75, i * 15), (5, 5), fill=color)\n",
    "    leg_element.add(rect)\n",
    "    text = svgwrite.text.Text(\n",
    "        desc, insert=(-70, i * 15 + 2.5), style=\"font-family:monospaced;font-size:2\"\n",
    "    )\n",
    "    leg_element.add(text)\n",
    "    legend_group.add(leg_element)\n",
    "dwg.add(legend_group)\n",
    "dwg.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reduce domain names - colors\n",
    "reduce_colors = {\n",
    "    i: [] for i in colormap.index\n",
    "}  # This dict should contain a list of domains that can be summed up into one\n",
    "for i, index in enumerate(legend.index):\n",
    "    ratios = {}\n",
    "    if type(index) == float or index == \"-\":\n",
    "        continue\n",
    "    for test_against in legend.index:\n",
    "        if (\n",
    "            type(test_against) == float\n",
    "            or test_against == \"-\"\n",
    "            or index == test_against\n",
    "            or len(test_against) < 5\n",
    "        ):\n",
    "            continue\n",
    "        ratio = fuzz.partial_ratio(index, test_against)\n",
    "        ratios[ratio] = test_against\n",
    "    best_ratio = max(ratios.keys())\n",
    "    best_match = ratios.get(best_ratio)\n",
    "    reduce_colors.get(best_match).append(index)\n",
    "\n",
    "reduce_colors = pandas.Series(reduce_colors)\n",
    "reduce_colors.to_csv(\"lib/color_reduction.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Cold-shock' DNA-binding domain                     []\n",
       "'Homeobox' domain profile.                          []\n",
       "'Homeobox' domain signature.                        []\n",
       "'Paired box' domain                                 []\n",
       "(E3-INDEPENDENT) E2 UBIQUITIN-CONJUGATING ENZYME    []\n",
       "                                                    ..\n",
       "znfxneu3                                            []\n",
       "zp_4                                                []\n",
       "zpr1                                                []\n",
       "zu_1                                                []\n",
       "zz_5                                                []\n",
       "Length: 27592, dtype: object"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reduce_colors"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.3 ('.venv': venv)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "vscode": {
   "interpreter": {
    "hash": "dc465123352b13523d34ae6bb9d107121edb3bfbca67b5c690ac65a480b324b1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
